### PR TITLE
Refactor shared package into the architecture

### PR DESCRIPTION
**Pull request by lenhard**
The new Java 9 module system brings us the possibility to actually modularize inside Java, i.e. to distribute our source code into multiple modules and specify dependencies and constraints among them, much in the way that we do it now with tests. But to extract modules, there can be no cyclic dependencies among them.

The `model` package is our best candidate for extracting a new module. Here the last dependency that coupled it to transitively to everything else was the `shared` package. This PR refactors the shared package and puts most of it into `logic`. I broke the dependency from `model` by introducing a number of interface that the `logic` layer implements. These interfaces will have just one implementation now, but at least they break the dependency cycle.

This is just a refactoring, so I think it should be merged after the release of 4.1

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [ ] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

File: src/main/java/org/jabref/gui/exporter/SaveDatabaseAction.java [hunk 1]
```diff
@@ -318,10 +320,13 @@ public void saveAs(File file) throws Exception {
 
         if (context.getLocation() == DatabaseLocation.SHARED) {
             // Save all properties dependent on the ID. This makes it possible to restore them.
-            DBMSConnectionProperties properties = context.getDBMSSynchronizer().getDBProcessor()
-                    .getDBMSConnectionProperties();
-            new SharedDatabasePreferences(context.getDatabase().generateSharedDatabaseID())
-                    .putAllDBMSConnectionProperties(properties);
+            DatabaseSynchronizer synchronizer = context.getDBMSSynchronizer();
+            if (synchronizer instanceof DBMSSynchronizer) {
+                DatabaseConnectionProperties properties = ((DBMSSynchronizer) synchronizer).getDBProcessor()
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/exporter/SaveDatabaseAction.java]: If you add `getDBMSConnectionProperties` to the `Synchronizer` interface, then you probably don't need the if statement and the explicit cast here.

Comment (lenhard) [reply src/main/java/org/jabref/gui/exporter/SaveDatabaseAction.java]: Thanks, nicely spotted! So far, I was focused on the refactoring only, but the code certainly has a lot of potential for improvements. Implementing your suggestion was fairly easy and I fixed this at a second position as well.

**Comment by koppor** [comment]
This refs https://github.com/JabRef/jabref/issues/110. I am still planning to work on that as I want to provide our logic package as standalone library.

**Comment by lenhard** [comment]
@koppor Ah, I was looking for that issue :) Maybe we should re-open it, since we now have a tangible use case (the new Java module system). Once the model package is modularized, the logic package will not be far off.

**Comment by lenhard** [comment]
@JabRef/developers With another review, this could be merged.

File: src/main/java/org/jabref/logic/shared/DBMSConnectionProperties.java [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
+import org.jabref.model.database.shared.DatabaseConnectionProperties;
 
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
 /**
  * Keeps all essential data for establishing a new connection to a DBMS using {@link DBMSConnection}.
  */
-public class DBMSConnectionProperties {
+public class DBMSConnectionProperties implements DatabaseConnectionProperties {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/shared/DBMSConnectionProperties.java]: I only don't get this, isn't this supposed to be same or are DataBaseConnectionProperties more general?

File: src/main/java/org/jabref/model/database/shared/DatabaseSynchronizer.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+    String getDBName();
+
+    void pullChanges();
+
+    void closeSharedDatabase();
+
+    void registerListener(Object listener);
+
+    void openSharedDatabase(DatabaseConnection connection) throws DatabaseNotSupportedException, SQLException;
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/database/shared/DatabaseSynchronizer.java]: I would let this rather be more general, by not having the concrete SQLException here. I could implement this interface for my Sharelatex Connection.

File: src/main/java/org/jabref/logic/shared/DBMSConnectionProperties.java (hunk 2 above)
Comment (lenhard) [reply src/main/java/org/jabref/logic/shared/DBMSConnectionProperties.java]: `DBMSConnectionProperties` are the only implementation of `DatabaseConnectionProperties`, they're essentially the same. The sole reason for the existence of the interface is that it breaks the dependency from model to elsewhere.

File: src/main/java/org/jabref/model/database/shared/DatabaseSynchronizer.java (hunk 3 above)
Comment (lenhard) [reply src/main/java/org/jabref/model/database/shared/DatabaseSynchronizer.java]: Ok, I've inlined the `SQLException` in favor of an `IllegalStateException`.

**Pull request merged**